lib/metrics.js

```diff
@@ -0,0 +1,848 @@
+   * @returns {string} The Dogstatsd-formatted metric report.
+   */
+  toDogStatsd() {
+    return `${this.toStatsd()}${this.#dogstatsdTags()}`;
+  }
+
+  /*
+   * Pack metadata into Dogstatsd-compatible tags.
+   *
+   * @returns {string} The packed metadata.
+   * @private
+   */
+  #dogstatsdTags() {
+    const entries = ObjectEntries(this.meta);
+    const pairs = ArrayPrototypeMap(entries, ({ 0: k, 1: v }) => `${k}:${v}`);
+    const tags = ArrayPrototypeJoin(pairs, ',');
+    return tags.length ? `|${tags}` : '';
+  }
+
+  /**
+   * Convert the metric report to a graphite-compatible string.
+   * @returns {string} The graphite-formatted metric report.
+   */
+  toGraphite() {
+    const { name, value, time } = this;
+    return `${name} ${value} ${MathFloor(time / 1000)}`;
+  }
+
+  /**
+   * Convert the metric report to a Prometheus-compatible string.
+   * @returns {string} The Prometheus-formatted metric report.
+   */
+  toPrometheus() {
+    const { name, value, time } = this;
+    return `${name}${this.#prometheusLabels()} ${value} ${time}`;
+  }
+
+  /*
+   * Pack metadata into Prometheus-compatible labels.
+   *
+   * @returns {string} The packed metadata.
+   * @private
+   */
+  #prometheusLabels() {
+    const entries = ObjectEntries(this.meta);
+    const pairs = ArrayPrototypeMap(entries, ({ 0: k, 1: v }) => `${k}="${v}"`);
+    const labels = ArrayPrototypeJoin(pairs, ',');
+    return labels.length ? `{${labels}}` : '';
+  }
+}
+
+/**
+ * Represents a metric which can be reported to.
+ */
+class Metric {
+  #channel;
+
+  /**
+   * The type of metric.
+   * @property {string} type
+   */
+
+  /**
+   * The name of the metric.
+   * @property {string} name
+   */
+
+  /**
+   * Additional metadata to include with the metric.
+   * @property {object} meta
+   */
+
+  /**
+   * Constructs a new metric.
+   * @param {string} type The type of metric.
+   * @param {string} name The name of the metric.
+   * @param {object} [meta] Additional metadata to include with the metric.
+   */
+  constructor(type, name, meta) {
+    if (!metricTypeNames.includes(type)) {
+      throw new ERR_INVALID_ARG_VALUE('type', type, wrongTypeErr);
+    }
+    if (typeof name !== 'string' || !name) {
+      throw new ERR_INVALID_ARG_TYPE('name', ['string'], name);
+    }
+    if (meta !== undefined && typeof meta !== 'object') {
+      throw new ERR_INVALID_ARG_TYPE('meta', ['object', 'undefined'], meta);
+    }
+
+    this.type = type;
+    this.name = name;
+    this.meta = meta;
+
+    // Before acquiring the channel, check if it already exists.
+    const exists = hasChannel(this.channelName);
+    this.#channel = channel(this.channelName);
+
+    // If the channel is new and there are new channel subscribers,
+    // publish the metric to the new metric channel.
+    if (!exists && newMetricChannel.hasSubscribers) {
+      newMetricChannel.publish(this);
+    }
+
+    ObjectFreeze(this);
```

Comment:
Why freezing?

Reply:
No reason, I just tend to freeze things when they don't need to be mutable. Prevents people accidentally messing them up. 🙈